File: android/app/src/androidTest/kotlin/com/menumaker/pageobjects/SellerCouponPage.kt
package com.menumaker.pageobjects

import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.ComposeTestRule

/**
 * Page Object for Seller Coupon Management Screen
 * Provides fluent API for seller coupon management interactions
 */
class SellerCouponPage(private val composeTestRule: ComposeTestRule) {

    // Elements
    private val createCouponButton = composeTestRule.onNode(
        hasText("create", substring = true, ignoreCase = true) or
        hasText("add coupon", substring = true, ignoreCase = true)
    )
    private val couponList = composeTestRule.onAllNodesWithTag("CouponItem")
    private val emptyStateMessage = composeTestRule.onNodeWithText("no coupon", substring = true, ignoreCase = true)

    // Coupon form elements
    private val couponCodeField = composeTestRule.onNodeWithTag("coupon-code-field")
    private val percentageButton = composeTestRule.onNode(
        hasText("percentage", substring = true, ignoreCase = true) or hasText("%", substring = true)
    )
    private val fixedAmountButton = composeTestRule.onNode(
        hasText("fixed", substring = true, ignoreCase = true) or hasText("₹", substring = true)
    )
    private val discountValueField = composeTestRule.onNodeWithTag("discount-value-field")
    private val minOrderAmountField = composeTestRule.onNodeWithTag("min-order-amount-field")
    private val maxDiscountField = composeTestRule.onNodeWithTag("max-discount-field")
    private val usageLimitField = composeTestRule.onNodeWithTag("usage-limit-field")
    private val saveCouponButton = composeTestRule.onNode(
        hasText("save", substring = true, ignoreCase = true) or
        hasText("create", substring = true, ignoreCase = true)
    )
    private val cancelButton = composeTestRule.onNodeWithText("cancel", substring = true, ignoreCase = true)
    private val toggleActiveSwitch = composeTestRule.onNodeWithTag("toggle-active-switch")
    private val deleteButton = composeTestRule.onNodeWithText("delete", substring = true, ignoreCase = true)
    private val confirmDeleteButton = composeTestRule.onNode(
        hasText("confirm", substring = true, ignoreCase = true) or
        hasText("yes", ignoreCase = true)
    )

    // Actions
    fun tapCreateCoupon(): SellerCouponPage {
        composeTestRule.onNode(
            hasTestTag("CreateCouponButton") or
            hasText("create", substring = true, ignoreCase = true) or
            hasText("add coupon", substring = true, ignoreCase = true) or
            hasText("new coupon", substring = true, ignoreCase = true)
        ).performClick()
        Thread.sleep(1000)
        return this
    }

    fun enterCouponCode(code: String): SellerCouponPage {
        composeTestRule.onNode(
            hasTestTag("coupon-code-field") or
            hasTestTag("CouponCodeField") or
            hasText("code", substring = true, ignoreCase = true)
        ).performTextInput(code)
        return this
    }

    fun selectDiscountType(type: DiscountType): SellerCouponPage {
        when (type) {
            DiscountType.PERCENTAGE -> composeTestRule.onNode(
                hasTestTag("PercentageDiscountButton") or
                hasText("percentage", substring = true, ignoreCase = true) or
                hasText("%", substring = true)
            ).performClick()
            DiscountType.FIXED_AMOUNT -> composeTestRule.onNode(
                hasTestTag("FixedAmountButton") or
                hasText("fixed", substring = true, ignoreCase = true) or
                hasText("₹", substring = true) or
                hasText("amount", substring = true, ignoreCase = true)
            ).performClick()
        }
        return this
    }

    fun enterDiscountValue(value: String): SellerCouponPage {
        composeTestRule.onNode(
            hasTestTag("discount-value-field") or
            hasTestTag("DiscountValueField") or
            hasText("value", substring = true, ignoreCase = true) or
            hasText("discount", substring = true, ignoreCase = true)
        ).performTextInput(value)
        return this
    }

    fun enterMinOrderAmount(amount: String): SellerCouponPage {
        try {
            composeTestRule.onNode(
                hasTestTag("min-order-amount-field") or
                hasTestTag("MinOrderAmountField") or
                hasText("minimum order", substring = true, ignoreCase = true) or
                hasText("min amount", substring = true, ignoreCase = true)
            ).performTextInput(amount)
        } catch (e: AssertionError) {
            // Field not found or not required
        }
        return this
    }

    fun enterMaxDiscount(amount: String): SellerCouponPage {
        try {
            composeTestRule.onNode(
                hasTestTag("max-discount-field") or
                hasTestTag("MaxDiscountField") or
                hasText("maximum discount", substring = true, ignoreCase = true) or
                hasText("max discount", substring = true, ignoreCase = true)
            ).performTextInput(amount)
        } catch (e: AssertionError) {
            // Field not found or not required
        }
        return this
    }

    fun enterUsageLimit(limit: String): SellerCouponPage {
        try {
            composeTestRule.onNode(
                hasTestTag("usage-limit-field") or
                hasTestTag("UsageLimitField") or
                hasText("usage limit", substring = true, ignoreCase = true) or
                hasText("max uses", substring = true, ignoreCase = true)
            ).performTextInput(limit)
        } catch (e: AssertionError) {
            // Field not found or not required
        }
        return this
    }

    fun saveCoupon(): SellerCouponPage {
        composeTestRule.onNode(
            hasTestTag("SaveCouponButton") or
            hasText("save", substring = true, ignoreCase = true) or
            hasText("create", substring = true, ignoreCase = true) or
            hasText("add", substring = true, ignoreCase = true)
        ).performClick()
        Thread.sleep(1000)
        return this
    }

    fun cancelCouponCreation(): SellerCouponPage {
        composeTestRule.onNode(
            hasTestTag("CancelButton") or
            hasText("cancel", substring = true, ignoreCase = true) or
            hasText("back", ignoreCase = true)
        ).performClick()
        return this
    }

    fun tapFirstCoupon(): SellerCouponPage {
        if (couponList.fetchSemanticsNodes().isNotEmpty()) {
            couponList[0].performClick()
            Thread.sleep(1000)
        }
        return this
    }

    fun toggleCouponActive(): SellerCouponPage {
        try {
            composeTestRule.onNode(
                hasTestTag("toggle-active-switch") or
                hasTestTag("CouponActiveToggle") or
                hasContentDescription("toggle active", substring = true, ignoreCase = true)
            ).performClick()
            Thread.sleep(1000)
        } catch (e: AssertionError) {
            // Toggle not found or not applicable
        }
        return this
    }

    fun deleteCoupon(): SellerCouponPage {
        composeTestRule.onNode(
            hasTestTag("DeleteCouponButton") or
            hasText("delete", substring = true, ignoreCase = true) or
            hasText("remove", substring = true, ignoreCase = true)
        ).performClick()

        try {
            composeTestRule.onNode(
                hasTestTag("ConfirmDeleteButton") or
                hasText("confirm", substring = true, ignoreCase = true) or
                hasText("yes", ignoreCase = true) or
                hasText("delete", ignoreCase = true)
            ).performClick()
        } catch (e: AssertionError) {
            // Confirm button not found - direct delete
        }
        Thread.sleep(1000)
        return this
    }

    fun swipeToDelete(index: Int = 0): SellerCouponPage {
        val coupons = composeTestRule.onAllNodesWithTag("CouponItem")
        if (coupons.fetchSemanticsNodes().size > index) {
            coupons[index].performTouchInput {
                swipeLeft()
            }
            Thread.sleep(500)

            val deleteAction = composeTestRule.onNode(
                hasTestTag("SwipeDeleteAction") or
                hasText("Delete", ignoreCase = true) or
                hasText("Remove", ignoreCase = true)
            )
            try {
                deleteAction.performClick()
                Thread.sleep(500)
            } catch (e: AssertionError) {
                // Delete action not found
            }
        }
        return this
    }

    fun createCoupon(code: String, type: DiscountType, value: String, minOrder: String? = null): SellerCouponPage {
        tapCreateCoupon()
        enterCouponCode(code)
        selectDiscountType(type)
        enterDiscountValue(value)

        if (minOrder != null) {
            enterMinOrderAmount(minOrder)
        }

        saveCoupon()
        return this
    }

    // Assertions
    fun assertScreenDisplayed(): SellerCouponPage {
        composeTestRule.waitUntil(timeoutMillis = 2000) {
            try { createCouponButton.assertExists(); true } catch (e: AssertionError) { false } ||
            couponList.fetchSemanticsNodes().isNotEmpty() ||
            try { emptyStateMessage.assertExists(); true } catch (e: AssertionError) { false }
        }
        return this
    }

    fun assertCouponFormDisplayed(): SellerCouponPage {
        composeTestRule.onNode(
            hasTestTag("coupon-code-field") or
            hasTestTag("CouponCodeField") or
            hasText("code", substring = true, ignoreCase = true)
        ).assertExists()
        return this
    }

    fun assertCouponExists(code: String): SellerCouponPage {
        composeTestRule.onNode(
            hasTestTag("CouponItem_$code") or
            hasText(code, substring = true)
        ).assertExists()
        return this
    }

    fun assertCouponSaved(): SellerCouponPage {
        Thread.sleep(1000)
        // Verify form is closed by checking code field doesn't exist
        composeTestRule.onNode(
            hasTestTag("coupon-code-field") or
            hasTestTag("CouponCodeField")
        ).assertDoesNotExist()
        return this
    }

    fun assertCouponCount(expectedCount: Int): SellerCouponPage {
        val coupons = composeTestRule.onAllNodesWithTag("CouponItem")
        val actualCount = coupons.fetchSemanticsNodes().size
        assert(actualCount == expectedCount) {
            "Expected $expectedCount coupons, but found $actualCount"
        }
        return this
    }

    fun assertEmptyState(): SellerCouponPage {
        composeTestRule.onNode(
            hasTestTag("EmptyStateMessage") or
            hasText("no coupon", substring = true, ignoreCase = true) or
            hasText("no coupons available", substring = true, ignoreCase = true) or
            hasText("create your first coupon", substring = true, ignoreCase = true)
        ).assertExists()
        return this
    }

    enum class DiscountType {
        PERCENTAGE,
        FIXED_AMOUNT
    }
}

File: ios/MenuMakerUITests/PageObjects/SellerCouponPage.swift
//
//  SellerCouponPage.swift
//  MenuMakerUITests
//
//  Page Object for Seller Coupon Management Screen
//

import XCTest

struct SellerCouponPage {
    let app: XCUIApplication

    // MARK: - Elements

    var createCouponButton: XCUIElement {
        app.buttons["add-coupon-button"]
    }

    var couponList: XCUIElementQuery {
        app.scrollViews.otherElements.matching(identifier: "CouponItem")
    }

    var firstCoupon: XCUIElement {
        couponList.firstMatch
    }

    var emptyStateMessage: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'no coupon'")).firstMatch
    }

    // Coupon form elements
    var couponCodeField: XCUIElement {
        let field = app.textFields["coupon-code-field"]
        return field.exists ? field : app.textFields.matching(identifier: "coupon-code-field").firstMatch
    }

    var discountTypeButtons: XCUIElementQuery {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'percentage' OR label CONTAINS[c] 'fixed'"))
    }

    var percentageButton: XCUIElement {
        let button = app.buttons["percentage-button"]
        return button.exists ? button : app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'percentage' OR label CONTAINS '%'")).firstMatch
    }

    var fixedAmountButton: XCUIElement {
        let button = app.buttons["fixed-amount-button"]
        return button.exists ? button : app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'fixed' OR label CONTAINS '₹'")).firstMatch
    }

    var discountValueField: XCUIElement {
        let field = app.textFields["discount-value-field"]
        return field.exists ? field : app.textFields.matching(identifier: "discount-value-field").firstMatch
    }

    var minOrderAmountField: XCUIElement {
        let field = app.textFields["min-order-field"]
        return field.exists ? field : app.textFields.matching(identifier: "min-order-field").firstMatch
    }

    var maxDiscountField: XCUIElement {
        let field = app.textFields["max-discount-field"]
        return field.exists ? field : app.textFields.matching(identifier: "max-discount-field").firstMatch
    }

    var usageLimitField: XCUIElement {
        let field = app.textFields["usage-limit-field"]
        return field.exists ? field : app.textFields.matching(identifier: "usage-limit-field").firstMatch
    }

    var startDatePicker: XCUIElement {
        app.datePickers.matching(NSPredicate(format: "identifier CONTAINS 'start'")).firstMatch
    }

    var endDatePicker: XCUIElement {
        app.datePickers.matching(NSPredicate(format: "identifier CONTAINS 'end'")).firstMatch
    }

    var saveCouponButton: XCUIElement {
        let button = app.buttons["save-coupon-button"]
        return button.exists ? button : app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'save' OR label CONTAINS[c] 'create'")).firstMatch
    }

    var cancelButton: XCUIElement {
        let button = app.buttons["cancel-button"]
        return button.exists ? button : app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'cancel'")).firstMatch
    }

    var toggleActiveSwitch: XCUIElement {
        let toggle = app.switches["coupon-active-toggle"]
        return toggle.exists ? toggle : app.switches.matching(NSPredicate(format: "identifier CONTAINS 'active' OR identifier CONTAINS 'enabled'")).firstMatch
    }

    var deleteButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'delete'")).firstMatch
    }

    var confirmDeleteButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'confirm' OR label CONTAINS[c] 'yes'")).firstMatch
    }

    // MARK: - Actions

    @discardableResult
    func tapCreateCoupon() -> SellerCouponPage {
        createCouponButton.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func enterCouponCode(_ code: String) -> SellerCouponPage {
        couponCodeField.tap()
        couponCodeField.typeText(code)
        return self
    }

    @discardableResult
    func selectDiscountType(_ type: DiscountType) -> SellerCouponPage {
        switch type {
        case .percentage:
            percentageButton.tap()
        case .fixedAmount:
            fixedAmountButton.tap()
        }
        return self
    }

    @discardableResult
    func enterDiscountValue(_ value: String) -> SellerCouponPage {
        dismissKeyboardIfNeeded()  // Dismiss keyboard from previous field
        discountValueField.tap()
        discountValueField.typeText(value)
        return self
    }

    @discardableResult
    func enterMinOrderAmount(_ amount: String) -> SellerCouponPage {
        dismissKeyboardIfNeeded()  // Dismiss keyboard from previous field
        if minOrderAmountField.waitForExistence(timeout: 1) {
            minOrderAmountField.tap()
            minOrderAmountField.typeText(amount)
        }
        return self
    }

    @discardableResult
    func enterMaxDiscount(_ amount: String) -> SellerCouponPage {
        dismissKeyboardIfNeeded()  // Dismiss keyboard from previous field
        if maxDiscountField.waitForExistence(timeout: 1) {
            maxDiscountField.tap()
            maxDiscountField.typeText(amount)
        }
        return self
    }

    @discardableResult
    func enterUsageLimit(_ limit: String) -> SellerCouponPage {
        dismissKeyboardIfNeeded()  // Dismiss keyboard from previous field
        if usageLimitField.waitForExistence(timeout: 1) {
            usageLimitField.tap()
            usageLimitField.typeText(limit)
        }
        return self
    }

    @discardableResult
    func saveCoupon() -> SellerCouponPage {
        dismissKeyboardIfNeeded()
        saveCouponButton.tap()
        return self
    }

    @discardableResult
    func cancelCouponCreation() -> SellerCouponPage {
        cancelButton.tap()
        return self
    }

    @discardableResult
    func tapFirstCoupon() -> SellerCouponPage {
        firstCoupon.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func toggleCouponActive() -> SellerCouponPage {
        if toggleActiveSwitch.waitForExistence(timeout: 1) {
            toggleActiveSwitch.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func deleteCoupon() -> SellerCouponPage {
        deleteButton.tap()
        if confirmDeleteButton.waitForExistence(timeout: 1) {
            confirmDeleteButton.tap()
        }
        sleep(1)
        return self
    }

    @discardableResult
    func swipeToDelete(at index: Int = 0) -> SellerCouponPage {
        let coupon = couponList.element(boundBy: index)
        coupon.swipeLeft()
        let deleteAction = app.buttons["Delete"]
        if deleteAction.waitForExistence(timeout: 1) {
            deleteAction.tap()
        }
        return self
    }

    func createCoupon(code: String, type: DiscountType, value: String, minOrder: String? = nil) {
        tapCreateCoupon()
        enterCouponCode(code)
        selectDiscountType(type)
        enterDiscountValue(value)

        if let minOrder = minOrder {
            enterMinOrderAmount(minOrder)
        }

        saveCoupon()
    }

    // MARK: - Assertions

    @discardableResult
    func assertScreenDisplayed(timeout: TimeInterval = 2) -> SellerCouponPage {
        XCTAssertTrue(createCouponButton.waitForExistence(timeout: timeout) ||
                     firstCoupon.waitForExistence(timeout: timeout) ||
                     emptyStateMessage.waitForExistence(timeout: timeout),
                     "Coupon management screen should be displayed")
        return self
    }

    @discardableResult
    func assertCouponFormDisplayed() -> SellerCouponPage {
        XCTAssertTrue(couponCodeField.waitForExistence(timeout: 2), "Coupon form should be displayed")
        return self
    }

    @discardableResult
    func assertCouponExists(_ code: String) -> SellerCouponPage {
        let couponLabel = app.staticTexts[code]
        XCTAssertTrue(couponLabel.exists, "Coupon '\(code)' should exist in the list")
        return self
    }

    @discardableResult
    func assertCouponSaved() -> SellerCouponPage {
        sleep(2)  // Wait longer for async coupon creation and form dismissal
        XCTAssertFalse(couponCodeField.exists, "Coupon form should be dismissed after save")
        return self
    }

    @discardableResult
    func assertCouponCount(_ expectedCount: Int) -> SellerCouponPage {
        let actualCount = couponList.count
        XCTAssertEqual(actualCount, expectedCount, "Should have \(expectedCount) coupons, found \(actualCount)")
        return self
    }

    @discardableResult
    func assertEmptyState() -> SellerCouponPage {
        XCTAssertTrue(emptyStateMessage.exists, "Empty state should be displayed")
        return self
    }

    // MARK: - Types

    enum DiscountType {
        case percentage
        case fixedAmount
    }

    // MARK: - Helpers

    private func dismissKeyboardIfNeeded() {
        if app.keyboards.count > 0 {
            // Try Done button first (from keyboard toolbar)
            let doneButton = app.toolbars.buttons["Done"]
            if doneButton.exists {
                doneButton.tap()
                return
            }

            // Try Return button for text keyboards
            let returnButton = app.keyboards.buttons["Return"]
            if returnButton.exists {
                returnButton.tap()
                return
            }

            // For numberPad/decimalPad without toolbar, tap outside to dismiss
            // Tap the navigation bar area
            let navBar = app.navigationBars.firstMatch
            if navBar.exists {
                navBar.tap()
            }
        }
    }
}
